FileSystemException on Windows with live-reload enabled

### Describe the bug

On Windows, with browser-live-reload enabled, the quarkus-web-bundler extension will try to create symlinks. Windows prohibits this for normal users.

The symlink is created here:
https://github.com/quarkiverse/quarkus-web-bundler/blob/1.5.3/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/PrepareForBundlingProcessor.java#L204


Please note that the exception message below is in Dutch, because that's how Windows passes it along. I've looked up what the error message would be in English:

* Dutch: Een van de vereiste bevoegdheden is niet aan de client toegekend
* English: A required privilege is not held by the client
* Windows error code: 1314

### Expected behavior

Successful start.

### Actual behavior

An exception occurs:

```
Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor#prepareForBundling threw an exception: java.io.UncheckedIOException: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
```

### How to Reproduce?

1. `quarkus.cmd create app -x quarkus-web-bundler -code`
2. `cd .\code-with-quarkus\`
3. `quarkus.cmd dev`
    * Exception: `java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: A required privilege is not held by the client` (message translated)
4. `echo 'quarkus.web-bundler.browser-live-reload=false' >> .\src\main\resources\application.properties
5. `quarkus.cmd dev`
    * Successful start

### Output of `uname -a` or `ver`

Windows 11 Pro (version 10.0.22631 N/A Build 22631)

### Output of `java -version`

openjdk version "21.0.2" 2024-01-16

### Quarkus version or git rev

Quarkus: 3.11.2
quarkus-web-bundler: 1.5.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

Full stack trace:

```
2024-06-17 03:26:52,995 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor#prepareForBundling threw an exception: java.io.UncheckedIOException: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
        at io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.prepareForBundling(PrepareForBundlingProcessor.java:259)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.createSymbolicLink(WindowsFileSystemProvider.java:604)
        at java.base/java.nio.file.Files.createSymbolicLink(Files.java:1069)
        at io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.prepareForBundling(PrepareForBundlingProcessor.java:204)
        ... 10 more

        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:252)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:58)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:113)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:435)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:56)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor#prepareForBundling threw an exception: java.io.UncheckedIOException: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
        at io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.prepareForBundling(PrepareForBundlingProcessor.java:259)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.createSymbolicLink(WindowsFileSystemProvider.java:604)
        at java.base/java.nio.file.Files.createSymbolicLink(Files.java:1069)
        at io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.prepareForBundling(PrepareForBundlingProcessor.java:204)
        ... 10 more

        at io.quarkus.builder.Execution.run(Execution.java:123)
        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:327)
        ... 9 more
Caused by: java.io.UncheckedIOException: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
        at io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.prepareForBundling(PrepareForBundlingProcessor.java:259)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.nio.file.FileSystemException: C:\Users\sjoer\Documents\proj\qdemo\code-with-quarkus\target\web-bundler\dev\app\app.js: Een van de vereiste bevoegdheden is niet aan de client toegekend
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.createSymbolicLink(WindowsFileSystemProvider.java:604)
        at java.base/java.nio.file.Files.createSymbolicLink(Files.java:1069)
        at io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.prepareForBundling(PrepareForBundlingProcessor.java:204)
        ... 10 more
```

--- a/common-deployment/src/main/java/io/quarkiverse/web/bundler/deployment/WebBundlerConfig.java
+++ b/common-deployment/src/main/java/io/quarkiverse/web/bundler/deployment/WebBundlerConfig.java
@@ -456,7 +456,6 @@ public interface WebBundlerConfig {
         }
 
         return Objects.equals(c1.webRoot(), c2.webRoot())
-                && Objects.equals(c1.bundle(), c2.bundle())
                 && Objects.equals(c1.staticDir(), c2.staticDir())
                 && Objects.equals(c1.bundlePath(), c2.bundlePath())
                 && BundlingConfig.isEqual(c1.bundling(), c2.bundling())

--- a/common-deployment/src/main/java/io/quarkiverse/web/bundler/deployment/items/BundleConfigAssetsBuildItem.java
+++ b/common-deployment/src/main/java/io/quarkiverse/web/bundler/deployment/items/BundleConfigAssetsBuildItem.java
@@ -2,6 +2,9 @@ package io.quarkiverse.web.bundler.deployment.items;
 
 import java.util.List;
 
+/**
+ * This contains config for bundling such as tsconfig.json
+ */
 public final class BundleConfigAssetsBuildItem extends WebAssetsBuildItem {
 
     public BundleConfigAssetsBuildItem(List<WebAsset> webAssets) {

--- a/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/BundlingProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/BundlingProcessor.java
@@ -46,6 +46,14 @@ class BundlingProcessor {
         if (readyForBundling == null) {
             return;
         }
+        bundleAndProcess(config, readyForBundling, staticResourceProducer, generatedBundleProducer,
+                generatedEntryPointProducer);
+    }
+
+    static BundleResult bundleAndProcess(WebBundlerConfig config, ReadyForBundlingBuildItem readyForBundling,
+            BuildProducer<GeneratedWebResourceBuildItem> staticResourceProducer,
+            BuildProducer<GeneratedBundleBuildItem> generatedBundleProducer,
+            BuildProducer<GeneratedEntryPointBuildItem> generatedEntryPointProducer) {
         try {
             final long startedBundling = Instant.now().toEpochMilli();
             final BundleResult result = Bundler.bundle(readyForBundling.bundleOptions(), false);
@@ -55,6 +63,7 @@ class BundlingProcessor {
 
             handleBundleDistDir(config, generatedBundleProducer, staticResourceProducer, result.dist(), startedBundling);
             processGeneratedEntryPoints(config, readyForBundling.bundleOptions().workDir(), generatedEntryPointProducer);
+            return result;
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }

--- a/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/DevModeBundlingProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/DevModeBundlingProcessor.java
@@ -1,15 +1,12 @@
 package io.quarkiverse.web.bundler.deployment;
 
-import static io.quarkiverse.web.bundler.deployment.BundlingProcessor.handleBundleDistDir;
-import static io.quarkiverse.web.bundler.deployment.BundlingProcessor.processGeneratedEntryPoints;
+import static io.quarkiverse.web.bundler.deployment.BundlingProcessor.*;
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.nio.file.Path;
-import java.time.Instant;
 import java.util.HashMap;
 import java.util.Set;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicReference;
 
 import org.jboss.logging.Logger;
@@ -18,6 +15,7 @@ import io.mvnpm.esbuild.BundleException;
 import io.mvnpm.esbuild.Bundler;
 import io.mvnpm.esbuild.Watch;
 import io.mvnpm.esbuild.model.BundleOptions;
+import io.mvnpm.esbuild.model.BundleResult;
 import io.quarkiverse.web.bundler.deployment.items.GeneratedBundleBuildItem;
 import io.quarkiverse.web.bundler.deployment.items.GeneratedEntryPointBuildItem;
 import io.quarkiverse.web.bundler.deployment.items.ReadyForBundlingBuildItem;
@@ -39,8 +37,6 @@ public class DevModeBundlingProcessor {
     private static final AtomicReference<Watch> watchRef = new AtomicReference<>();
 
     private static final AtomicReference<BundleException> bundleExceptionRef = new AtomicReference<>();
-    private static final AtomicReference<CountDownLatch> WAITER = new AtomicReference<>();
-    private static volatile long lastBundling = 0;
 
     @BuildStep(onlyIf = IsDevelopment.class)
     void watch(WebBundlerConfig config,
@@ -57,38 +53,38 @@ public class DevModeBundlingProcessor {
         final BundlesBuildContext bundlesBuildContext = liveReload.getContextObject(BundlesBuildContext.class);
         final boolean isLiveReload = liveReload.isLiveReload();
         Watch watch = DevModeBundlingProcessor.watchRef.get();
-        if (isLiveReload && devService != null) {
-            boolean shouldShutdownTheBroker = bundlesBuildContext == null
-                    || watch == null
-                    || !watch.isAlive()
-                    || !WebBundlerConfig.isEqual(config, bundlesBuildContext.config());
-            if (!shouldShutdownTheBroker) {
-                try {
-                    if (readyForBundling.started() == null) {
-                        watch.updateEntries(readyForBundling.bundleOptions().entries());
-                        // We should just wait for the change to happen
-                        waitForBundling(readyForBundling);
-                    }
+        if (readyForBundling.started() == null) {
+            // no changes
+            boolean isRestartWatchNeeded = readyForBundling.enabledBundlingWatch() && (watch == null || !watch.isAlive());
+            if (!isRestartWatchNeeded) {
+                if (watch != null && watch.isAlive()) {
                     devServices.produce(devService.toBuildItem());
-                    final BundlesBuildContext newBundlesBuildContext = new BundlesBuildContext(readyForBundling.bundleOptions(),
-                            config, watch.dist());
-                    liveReload.setContextObject(BundlesBuildContext.class, newBundlesBuildContext);
-                    handleBundleDistDir(config, generatedBundleProducer, staticResourceProducer, watch.dist(),
-                            readyForBundling.started());
-                    processGeneratedEntryPoints(config, readyForBundling.bundleOptions().workDir(),
-                            generatedEntryPointProducer);
-                } catch (IOException e) {
-                    shutdownDevService();
-                    liveReload.setContextObject(BundlesBuildContext.class, new BundlesBuildContext());
-                    throw new UncheckedIOException(e);
-                } catch (Exception e) {
-                    shutdownDevService();
-                    liveReload.setContextObject(BundlesBuildContext.class, new BundlesBuildContext());
-                    throw e;
                 }
+                final BundlesBuildContext newBundlesBuildContext = new BundlesBuildContext(readyForBundling.bundleOptions(),
+                        bundlesBuildContext.bundleDistDir());
+                liveReload.setContextObject(BundlesBuildContext.class, newBundlesBuildContext);
+                handleBundleDistDir(config, generatedBundleProducer, staticResourceProducer,
+                        bundlesBuildContext.bundleDistDir(),
+                        readyForBundling.started());
+                processGeneratedEntryPoints(config, readyForBundling.bundleOptions().workDir(),
+                        generatedEntryPointProducer);
                 return;
             }
+        }
+
+        if (watch != null) {
             shutdownDevService();
+        }
+
+        if (!readyForBundling.enabledBundlingWatch()) {
+            // We use normal bundling when watch is not enabled
+            final BundleResult bundleResult = bundleAndProcess(config, readyForBundling, staticResourceProducer,
+                    generatedBundleProducer,
+                    generatedEntryPointProducer);
+            final BundlesBuildContext newBundlesBuildContext = new BundlesBuildContext(readyForBundling.bundleOptions(),
+                    bundleResult.dist());
+            liveReload.setContextObject(BundlesBuildContext.class, newBundlesBuildContext);
+            return;
         }
 
         if (!isLiveReload) {
@@ -108,7 +104,6 @@ public class DevModeBundlingProcessor {
                     return;
                 }
                 LOGGER.debugf("New bundling event received: %s", r);
-                lastBundling = Instant.now().toEpochMilli();
                 if (!r.isSuccess()) {
                     bundleExceptionRef.set(r.bundleException());
                     RuntimeUpdatesProcessor.INSTANCE.setRemoteProblem(r.bundleException());
@@ -120,10 +115,6 @@ public class DevModeBundlingProcessor {
                     shutdownDevService();
                 }
                 callNoRestartChangesConsumers(r.isSuccess());
-                final CountDownLatch countDownLatch = WAITER.get();
-                if (countDownLatch != null) {
-                    countDownLatch.countDown();
-                }
             }, false);
             watchRef.set(watch);
             devService = new DevServicesResultBuildItem.RunningDevService(
@@ -133,7 +124,7 @@ public class DevModeBundlingProcessor {
                 throw watch.firstBuildResult().bundleException();
             }
             final BundlesBuildContext newBundlesBuildContext = new BundlesBuildContext(readyForBundling.bundleOptions(),
-                    config, watch.dist());
+                    watch.dist());
             liveReload.setContextObject(BundlesBuildContext.class, newBundlesBuildContext);
             handleBundleDistDir(config, generatedBundleProducer, staticResourceProducer, watch.dist(),
                     readyForBundling.started());
@@ -176,46 +167,11 @@ public class DevModeBundlingProcessor {
         }
     }
 
-    private void waitForBundling(ReadyForBundlingBuildItem readyForBundling) {
-        if (readyForBundling.started() != null) {
-            if (lastBundling > readyForBundling.started()) {
-                LOGGER.debug("Bundling done, no need to wait");
-            } else {
-                final CountDownLatch latch = new CountDownLatch(1);
-                final CountDownLatch existingLatch = WAITER.getAndSet(latch);
-                WAITER.set(latch);
-                LOGGER.info("Bundling started...");
-                try {
-                    latch.await();
-                    LOGGER.debug("Bundling completed!");
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    shutdownDevService();
-                    throw new RuntimeException(e);
-                } finally {
-                    if (existingLatch != null) {
-                        // this will always countdown the previous latch,
-                        // it has no effect if it was already done
-                        existingLatch.countDown();
-                    }
-                }
-            }
-        }
-
-        final BundleException bundleException = bundleExceptionRef.get();
-        if (bundleException != null) {
-            shutdownDevService();
-            throw bundleException;
-        }
-
-    }
-
     record BundlesBuildContext(BundleOptions bundleOptions,
-            WebBundlerConfig config,
             Path bundleDistDir) {
 
         public BundlesBuildContext() {
-            this(null, null, null);
+            this(null, null);
         }
     }
 }

--- a/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/PrepareForBundlingProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/PrepareForBundlingProcessor.java
@@ -40,6 +40,7 @@ import io.quarkus.vertx.http.runtime.HttpBuildTimeConfig;
 public class PrepareForBundlingProcessor {
 
     private static final Logger LOGGER = Logger.getLogger(PrepareForBundlingProcessor.class);
+    public static volatile boolean enableBundlingWatch = true;
 
     private static final Map<EsBuildConfig.Loader, Function<LoadersConfig, Optional<Set<String>>>> LOADER_CONFIGS = Map
             .ofEntries(
@@ -120,20 +121,27 @@ public class PrepareForBundlingProcessor {
                 && WebBundlerConfig.isEqual(config, prepareForBundlingContext.config())
                 && Objects.equals(installedWebDependencies.list(), prepareForBundlingContext.dependencies())
                 && !liveReload.getChangedResources().contains(config.fromWebRoot("tsconfig.json"))
-                && entryPoints.equals(prepareForBundlingContext.entryPoints())) {
-            // We need to set non-restart watched file again
-            for (EntryPointBuildItem entryPoint : entryPoints) {
-                for (BundleWebAsset webAsset : entryPoint.getWebAssets()) {
-                    if (webAsset.isFile() && config.browserLiveReload()) {
-                        watchedFiles.produce(HotDeploymentWatchedFileBuildItem.builder()
-                                .setRestartNeeded(webAsset.srcFilePath().isEmpty())
-                                .setLocation(webAsset.resourceName())
-                                .build());
+                && entryPoints.equals(prepareForBundlingContext.entryPoints())
+                && entryPoints.stream().map(EntryPointBuildItem::getWebAssets).flatMap(List::stream)
+                        .map(WebAsset::resourceName)
+                        .noneMatch(liveReload.getChangedResources()::contains)) {
+            if (config.browserLiveReload() && enableBundlingWatch) {
+                // We need to set non-restart watched file again
+                for (EntryPointBuildItem entryPoint : entryPoints) {
+                    for (BundleWebAsset webAsset : entryPoint.getWebAssets()) {
+                        if (webAsset.isFile() && webAsset.srcFilePath().isPresent()) {
+                            watchedFiles.produce(HotDeploymentWatchedFileBuildItem.builder()
+                                    .setRestartNeeded(false)
+                                    .setLocation(webAsset.resourceName())
+                                    .build());
+                        }
                     }
                 }
             }
-            return new ReadyForBundlingBuildItem(prepareForBundlingContext.bundleOptions(), null, targetDir.dist());
+            return new ReadyForBundlingBuildItem(prepareForBundlingContext.bundleOptions(), null, targetDir.dist(),
+                    enableBundlingWatch);
         }
+        enableBundlingWatch = true;
 
         try {
             Files.createDirectories(targetDir.webBundler());
@@ -237,7 +245,7 @@ public class PrepareForBundlingProcessor {
             final BundleOptions options = optionsBuilder.build();
             liveReload.setContextObject(PrepareForBundlingContext.class,
                     new PrepareForBundlingContext(config, installedWebDependencies.list(), entryPoints, options));
-            return new ReadyForBundlingBuildItem(options, started, targetDir.dist());
+            return new ReadyForBundlingBuildItem(options, started, targetDir.dist(), enableBundlingWatch);
         } catch (IOException e) {
             liveReload.setContextObject(PrepareForBundlingContext.class, new PrepareForBundlingContext());
             throw new UncheckedIOException(e);
@@ -251,22 +259,39 @@ public class PrepareForBundlingProcessor {
             Files.write(targetPath, webAsset.resource().content(), StandardOpenOption.CREATE,
                     StandardOpenOption.TRUNCATE_EXISTING);
         } else {
-            if (browserLiveReload) {
-                createSymbolicLink(watchedFiles, webAsset, targetPath);
+            if (browserLiveReload && enableBundlingWatch) {
+                createSymbolicLinkOrFallback(watchedFiles, webAsset, targetPath);
             } else {
                 Files.copy(webAsset.resource().path(), targetPath, StandardCopyOption.REPLACE_EXISTING);
             }
         }
     }
 
-    static void createSymbolicLink(BuildProducer<HotDeploymentWatchedFileBuildItem> watchedFiles, WebAsset webAsset,
+    static void createSymbolicLinkOrFallback(BuildProducer<HotDeploymentWatchedFileBuildItem> watchedFiles, WebAsset webAsset,
             Path targetPath) throws IOException {
         Files.deleteIfExists(targetPath);
-        watchedFiles.produce(HotDeploymentWatchedFileBuildItem.builder()
-                .setRestartNeeded(webAsset.srcFilePath().isEmpty())
-                .setLocation(webAsset.resourceName())
-                .build());
-        Files.createSymbolicLink(targetPath, webAsset.srcFilePath().orElse(webAsset.resource().path()));
+        final boolean srcDetected = webAsset.srcFilePath().isPresent();
+        if (srcDetected) {
+            try {
+                Files.createSymbolicLink(targetPath, webAsset.srcFilePath().get());
+                // the default is restart for all web files, let's disable restart for this one.
+                // it will be detected by esbuild watcher
+                watchedFiles.produce(HotDeploymentWatchedFileBuildItem.builder()
+                        .setRestartNeeded(false)
+                        .setLocation(webAsset.resourceName())
+                        .build());
+            } catch (UnsupportedOperationException e) {
+                enableBundlingWatch = false;
+                LOGGER.warn(
+                        "Creating a symbolic link was not authorized on this system. It is required by the Web Bundler to allow filesystem watch. As a result, Web Bundler live-reload will use a scheduler as a fallback.\n\nTo resolve this issue, please add the necessary permissions to allow symbolic link creation.");
+                Files.copy(webAsset.resource().path(), targetPath, StandardCopyOption.REPLACE_EXISTING);
+            }
+        } else {
+            LOGGER.warn(
+                    "The sources are necessary by the Web Bundler to allow filesystem watch. Web Bundler live-reload will use a scheduler as a fallback");
+            enableBundlingWatch = false;
+            Files.copy(webAsset.resource().path(), targetPath, StandardCopyOption.REPLACE_EXISTING);
+        }
     }
 
     private byte[] readLiveReloadJs() throws IOException {

--- a/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/StaticWebAssetsProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/StaticWebAssetsProcessor.java
@@ -1,6 +1,6 @@
 package io.quarkiverse.web.bundler.deployment;
 
-import static io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.createSymbolicLink;
+import static io.quarkiverse.web.bundler.deployment.PrepareForBundlingProcessor.createSymbolicLinkOrFallback;
 import static io.quarkiverse.web.bundler.deployment.util.PathUtils.prefixWithSlash;
 
 import java.io.IOException;
@@ -39,7 +39,7 @@ public class StaticWebAssetsProcessor {
                 } else {
                     if (browserLiveReload) {
                         Files.createDirectories(targetPath.getParent());
-                        createSymbolicLink(watchedFileBuildItemProducer, webAsset, targetPath);
+                        createSymbolicLinkOrFallback(watchedFileBuildItemProducer, webAsset, targetPath);
                         makePublic(staticResourceProducer, prefixWithSlash(publicPath), targetPath, SourceType.STATIC_ASSET);
                     } else {
                         // We can read the file

--- a/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/items/ReadyForBundlingBuildItem.java
+++ b/deployment/src/main/java/io/quarkiverse/web/bundler/deployment/items/ReadyForBundlingBuildItem.java
@@ -13,10 +13,13 @@ public final class ReadyForBundlingBuildItem extends SimpleBuildItem {
 
     private final Path distDir;
 
-    public ReadyForBundlingBuildItem(BundleOptions bundleOptions, Long started, Path distDir) {
+    private final boolean disableBundlingWatch;
+
+    public ReadyForBundlingBuildItem(BundleOptions bundleOptions, Long started, Path distDir, boolean disableBundlingWatch) {
         this.bundleOptions = bundleOptions;
         this.started = started;
         this.distDir = distDir;
+        this.disableBundlingWatch = disableBundlingWatch;
     }
 
     public Long started() {
@@ -29,5 +32,9 @@ public final class ReadyForBundlingBuildItem extends SimpleBuildItem {
 
     public Path distDir() {
         return distDir;
+    }
+
+    public boolean enabledBundlingWatch() {
+        return disableBundlingWatch;
     }
 }

--- a/deployment/src/test/java/io/quarkiverse/web/bundler/test/WebBundlerDevModeTest.java
+++ b/deployment/src/test/java/io/quarkiverse/web/bundler/test/WebBundlerDevModeTest.java
@@ -45,7 +45,6 @@ public class WebBundlerDevModeTest {
                 .body(Matchers.containsString("console.log(\"Hello World! Modified!\");"));
         test.modifyResourceFile("web/app/app.css", s -> s.replace("background-color: #6b6bf5;", "background-color: #123456;"));
         test.modifyResourceFile("web/app/other.scss", s -> s.replace("color: #AAAAAA;", "color: #567890;"));
-        Thread.sleep(1000);
         RestAssured.given()
                 .get("/foo/bar/static/bundle/main.css")
                 .then()

--- a/docs/modules/ROOT/pages/includes/attributes.adoc
+++ b/docs/modules/ROOT/pages/includes/attributes.adoc
@@ -1,4 +1,4 @@
-:project-version: 1.6.0
+:project-version: 1.6.1
 
 :maven-version: 3.8.1+
 

--- a/runtime/src/main/java/io/quarkiverse/web/bundler/runtime/devmode/WebBundlerHotReplacementSetup.java
+++ b/runtime/src/main/java/io/quarkiverse/web/bundler/runtime/devmode/WebBundlerHotReplacementSetup.java
@@ -34,16 +34,24 @@ public class WebBundlerHotReplacementSetup implements HotReplacementSetup {
 
     private void startWatchScheduler() {
         if (scheduler == null) {
+            context.addPreScanStep(() -> {
+                // Make sure we don't scan when another scan is running
+                if (nextUpdate != Long.MAX_VALUE) {
+                    nextUpdate = System.currentTimeMillis() + 1000;
+                }
+            });
             scheduler = EXECUTOR.scheduleAtFixedRate(() -> {
                 try {
                     if (context.getDeploymentProblem() == null) {
                         // Let's not scan when there is a deployment problem
-                        // and wait for another request to trigger the scan
+                        // and wait for another request to trigger the scan`
                         if (System.currentTimeMillis() > nextUpdate) {
+                            // Make sure we don't call scan more than once
+                            nextUpdate = Long.MAX_VALUE;
                             if (context.doScan(false)) {
-                                LOGGER.debug("App restarted from watcher, let's wait 5s before watching again");
-                                // If we have restarted, let's wait 5s before another scan
-                                nextUpdate = System.currentTimeMillis() + 5000;
+                                LOGGER.debug("App restarted from watcher, let's wait 3s before watching again");
+                                // If we have restarted, let's wait 3s before another scan
+                                nextUpdate = System.currentTimeMillis() + 3000;
                             } else {
                                 nextUpdate = System.currentTimeMillis() + 1000;
                             }
@@ -51,7 +59,8 @@ public class WebBundlerHotReplacementSetup implements HotReplacementSetup {
                     }
 
                 } catch (Exception e) {
-                    throw new IllegalStateException(e);
+                    nextUpdate = System.currentTimeMillis() + 3000;
+                    throw new RuntimeException(e);
                 }
             }, 500, 500, TimeUnit.MILLISECONDS);
         }
@@ -62,12 +71,13 @@ public class WebBundlerHotReplacementSetup implements HotReplacementSetup {
     public void close() {
         HotReplacementSetup.super.close();
         if (scheduler != null) {
-            scheduler.cancel(false);
+            scheduler.cancel(true);
             scheduler = null;
         }
     }
 
     private void noRestartChanges(Set<String> strings) {
+        nextUpdate = System.currentTimeMillis() + 1000;
         for (Consumer<Set<String>> changeEventListener : changeEventListeners) {
             changeEventListener.accept(strings);
         }

